{"cells":[{"cell_type":"code","source":["!pip install biopython"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"0wWvxTUMHP1s","executionInfo":{"status":"ok","timestamp":1653329457327,"user_tz":-180,"elapsed":5672,"user":{"displayName":"Svetlana Milrud","userId":"08332842915222559816"}},"outputId":"3f7db6d2-2030-4a74-c2ac-a50f990eca3a"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n","Collecting biopython\n","  Downloading biopython-1.79-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.whl (2.3 MB)\n","\u001b[K     |████████████████████████████████| 2.3 MB 4.8 MB/s \n","\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from biopython) (1.21.6)\n","Installing collected packages: biopython\n","Successfully installed biopython-1.79\n"]}]},{"cell_type":"markdown","source":["## Import data (prot.fasta)"],"metadata":{"id":"97y-SVgyUS6F"}},{"cell_type":"code","source":["fasta_file=\"/content/prot.fasta\""],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Wwv9pq5YNr3d","executionInfo":{"status":"ok","timestamp":1653232250851,"user_tz":-180,"elapsed":280,"user":{"displayName":"Svetlana Milrud","userId":"08332842915222559816"}},"outputId":"385e505d-a918-4c74-f3f4-872bfe07a60e"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["/content/prot.fasta\n"]}]},{"cell_type":"markdown","source":["## Task 1\n","\n","Let's remeber how to use dictionaries.\n","\n","Task: return a dictionary where \n"," * keys are IDs of seqs from an input fasta file (prot.fasta),\n"," * key's values are seqs itself. "],"metadata":{"id":"dnSHoBOTUCgg"}},{"cell_type":"code","source":["def my_own_fasta_parser(inFile):\n","  sequences = {}\n","  with open (inFile) as handle:\n","      for f in handle:\n","          line = f.replace('\\n','')\n","          if line.startswith(\">\"):\n","              seq_id = line[1:]\n","          else:\n","              sequences[seq_id]=line\n","\n","  return sequences\n","\n","my_own_fasta_parser(fasta_file)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"bi0xgU8OIP9J","executionInfo":{"status":"ok","timestamp":1653231935844,"user_tz":-180,"elapsed":292,"user":{"displayName":"Svetlana Milrud","userId":"08332842915222559816"}},"outputId":"3d1bc30e-874b-4997-82d5-e04f1520a2e2"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["{'seq0': 'FQTWEEFSRAAEKLYLADPMKVRVVLKYRHVDGNLCIKVTDDLVCLVYRTDQAQDVKKIEKF',\n"," 'seq1': 'KYRTWEEFTRAAEKLYQADPMKVRVVLKYRHCDGNLCIKVTDDVVCLLYRTDQAQDVKKIEKFHSQLMRLME LKVTDNKECLKFKTDQAQEAKKMEKLNNIFFTLM',\n"," 'seq10': 'FDSWDEFVSKSVELFRNHPDTTRYVVKYRHCEGKLVLKVTDNHECLKFKTDQAQDAKKMEK',\n"," 'seq2': 'EEYQTWEEFARAAEKLYLTDPMKVRVVLKYRHCDGNLCMKVTDDAVCLQYKTDQAQDVKKVEKLHGK',\n"," 'seq3': 'MYQVWEEFSRAVEKLYLTDPMKVRVVLKYRHCDGNLCIKVTDNSVCLQYKTDQAQDVK',\n"," 'seq4': 'EEFSRAVEKLYLTDPMKVRVVLKYRHCDGNLCIKVTDNSVVSYEMRLFGVQKDNFALEHSLL',\n"," 'seq5': 'SWEEFAKAAEVLYLEDPMKCRMCTKYRHVDHKLVVKLTDNHTVLKYVTDMAQDVKKIEKLTTLLMR',\n"," 'seq6': 'FTNWEEFAKAAERLHSANPEKCRFVTKYNHTKGELVLKLTDDVVCLQYSTNQLQDVKKLEKLSSTLLRSI',\n"," 'seq7': 'SWEEFVERSVQLFRGDPNATRYVMKYRHCEGKLVLKVTDDRECLKFKTDQAQDAKKMEKLNNIFF',\n"," 'seq8': 'SWDEFVDRSVQLFRADPESTRYVMKYRHCDGKLVLKVTDNKECLKFKTDQAQEAKKMEKLNNIFFTLM',\n"," 'seq9': 'KNWEDFEIAAENMYMANPQNCRYTMKYVHSKGHILLKMSDNVKCVQYRAENMPDLKK'}"]},"metadata":{},"execution_count":19}]},{"cell_type":"markdown","source":["## Task 2\n","\n","Another super easy task.\n","\n","We have the same fasta file (prot.fasta).\n","\n","Now we want to get a list with the ids of protein seqs that have \n","a relative frequency higher than a given threshold for a given residue.\n","\n","And don't forget to use my_own_fasta_parser function from a previous task!"],"metadata":{"id":"lxQqrGCoUo24"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"6nqnGJIJTg3j"},"outputs":[],"source":["def my_own_residue_abundance(input_file, residue, threshold=0.2):\n","    \n","    seq_id = []\n","    sequences = my_own_fasta_parser(input_file)\n","\n","    for seq_id, sequence in sequences.items():\n","        #<<<<<<<<<<...>>>>>>>>>>>\n","        #<<<<<<<<<<...>>>>>>>>>>>\n","            #<<<<<<<<<<...>>>>>>>>>>>\n","\n","    return seq_id"]},{"cell_type":"code","source":["def my_own_residue_abundance(input_file, residue, threshold=0.2):   \n","    sequence_id = []\n","    sequences = my_own_fasta_parser(input_file)\n","\n","    for seq_id, sequence in sequences.items():\n","        frequency=(sequence.count(residue))/len(sequence)\n","        if frequency > threshold:\n","          sequence_id.append(seq_id)\n","    return sequence_id"],"metadata":{"id":"d6NYTRQtU3Ec"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["print(my_own_residue_abundance(fasta_file, 'D', threshold=0.08))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"0SCkGbsaXEBs","executionInfo":{"status":"ok","timestamp":1653234188308,"user_tz":-180,"elapsed":5,"user":{"displayName":"Svetlana Milrud","userId":"08332842915222559816"}},"outputId":"2eba539a-710a-4e89-94e7-cf09a36ef7bc"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["['seq0', 'seq2', 'seq3', 'seq8', 'seq10']\n"]}]},{"cell_type":"code","source":["print(my_own_residue_abundance(fasta_file, 'Q', threshold=0.05))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"bXNy2O_scM7z","executionInfo":{"status":"ok","timestamp":1653234290364,"user_tz":-180,"elapsed":309,"user":{"displayName":"Svetlana Milrud","userId":"08332842915222559816"}},"outputId":"0003b10c-79f0-487f-f81d-f888d2eab308"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["['seq1', 'seq2', 'seq3']\n"]}]},{"cell_type":"markdown","source":["## Task 3\n","\n","Let's practice in Bio package using.\n","\n","Task:\n","1. read fasta file containing several DNA seqs (nucl.fasta)\n","2. subset seqs that have GC content > 45 and coding protein with aromaticity > 0.01\n","3. write a new fasta file with those protein(!) seqs\n","4. return the percentage of seqs that passed your filter\n","Hint: Bio.SeqIO, Bio.SeqRecord, Bio.SeqUtils"],"metadata":{"id":"wiyuXxfac4T1"}},{"cell_type":"code","source":["import Bio\n","from Bio import SeqIO\n","from Bio import SeqRecord\n","from Bio import SeqUtils"],"metadata":{"id":"3n7hvdBOGKie"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["fasta_file1=\"/content/nucl.fasta\""],"metadata":{"id":"br4XXwM7fkgz"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["from Bio.SeqUtils import GC\n","from Bio.SeqUtils.ProtParam import ProteinAnalysis\n","from Bio.SeqRecord import SeqRecord\n","from Bio.Seq import Seq\n","\n","\n","def my_own_filtering(input_file, output_file, filt_gc=45, filt_arom=0.125):\n","  sequences = {}\n","  c = 0 \n","  with open(input_file, \"r\") as content:   \n","      for record in SeqIO.parse(content, \"fasta\"):\n","        # calculate GC content using Bio\n","        c+=1\n","        calc_gc=GC(record.seq)\n","        # calculate aromaticity using Bio\n","        protein=str((record.seq).translate())\n","        line=ProteinAnalysis(protein)\n","        calc_arom=line.aromaticity()\n","        if calc_gc >= filt_gc and calc_arom >= filt_arom:\n","          sequences[record.id]=protein\n","  records = []\n","  for seq_id, seq in sequences.items():\n","    my_record=SeqRecord(Seq(seq), id=seq_id, name=seq_id, description=seq_id)\n","    records.append(my_record)\n","  SeqIO.write(records,output_file, \"fasta\")\n","  return(round(float(len(sequences)/c)*100))\n"],"metadata":{"id":"GuNY9-HIlTGz"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["my_own_filtering(fasta_file1, \"/content/check1.fasta\", filt_gc=45, filt_arom=0.125)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"yaFSvBYXzcZ4","executionInfo":{"status":"ok","timestamp":1653254774420,"user_tz":-180,"elapsed":373,"user":{"displayName":"Svetlana Milrud","userId":"08332842915222559816"}},"outputId":"0454cdfc-2650-4bb8-bcc6-f19d25c3f662"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/Bio/Seq.py:2983: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n","  BiopythonWarning,\n"]},{"output_type":"execute_result","data":{"text/plain":["14"]},"metadata":{},"execution_count":247}]},{"cell_type":"code","source":["for seq_record in SeqIO.parse(\"/content/check1.fasta\",\"fasta\"):\n","    print (seq_record)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"fnWxzZ67yGCY","executionInfo":{"status":"ok","timestamp":1653254783316,"user_tz":-180,"elapsed":282,"user":{"displayName":"Svetlana Milrud","userId":"08332842915222559816"}},"outputId":"f5581551-489c-4671-de70-56553a2cc837"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["ID: id5\n","Name: id5\n","Description: id5\n","Number of features: 0\n","Seq('ISFARPRH')\n"]}]},{"cell_type":"markdown","source":["## Task 4\n","\n","Continue practicing in Bio package using:\n","\n","Task:\n","\n","Complete the following code that should be able to return \n","the best alignment of two amino acid seqs (pairwise2.align.globalds) based on BLOSUM62 matrix from Bio.SubsMat.MatrixInfo.\n","http://rosalind.info/glossary/blosum62/"],"metadata":{"id":"k2iRuob1FTae"}},{"cell_type":"code","source":["from Bio import pairwise2\n","from Bio.SubsMat import MatrixInfo as matlist\n","def balign(first_seq, second_seq, gap_open, gap_extend):\n","  # Load the matrix\n","  matrix = matlist.blosum62\n","  # Generate the alignments\n","  alns = pairwise2.align.globalds(first_seq,second_seq,matrix, gap_open, gap_extend)\n","  # Extract the best alignment (first one in the alns list)\n","  top_aln = alns[0]\n","  # Print the alignment\n","  aln_A, aln_B, score, begin, end = top_aln.seqA,top_aln.seqB,top_aln.score,top_aln.start,top_aln.end\n","  return(f\"Best alignment: seq1 = {aln_A}, seq2 = { aln_B}, score = {round(score)}, begin = {begin},end = {end}\")\n","\n","\n"],"metadata":{"id":"HEzW3xg0o1qb"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["first_seq=\"MEEPQSDPSVEPPLSQETFSD\"\n","second_seq=\"MEESQSDISLELPLSQETFSGLW\"\n","gap_open=-1\n","gap_extend=-0.5\n","\n","balign(first_seq, second_seq, gap_open, gap_extend)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":36},"id":"iQ3rYr58pQxS","executionInfo":{"status":"ok","timestamp":1653254597937,"user_tz":-180,"elapsed":288,"user":{"displayName":"Svetlana Milrud","userId":"08332842915222559816"}},"outputId":"13ef8ddf-35d8-4139-fb0a-82abe1950d9e"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["'Best alignment: seq1 = MEEPQSDP-SVEP-PLSQETFSD--, seq2 = MEESQSD-ISLE-LPLSQETFSGLW, score = 72, begin = 0,end = 25'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{},"execution_count":245}]},{"cell_type":"markdown","source":["## Task 5\n","\n","You have some DNA sequence: *AGTACTAGAGCATTCTATGGAG*.\n","Find out which peptides could be created from it and sort them by their length.\n","Use as much Biopython modules as possible."],"metadata":{"id":"1UFa0-JsgaJ0"}},{"cell_type":"code","source":["\n","from Bio.Seq import Seq\n","from Bio.Data import CodonTable\n","\n","standard_table = CodonTable.unambiguous_dna_by_name[\"Standard\"]\n","protein_pep = []\n","seq = Seq(\"AGTACTAGAGCATTCTATGGAG\")\n","\n","# start from different nucleotide\n","for i in range (0,len(seq)-2):\n","  peptide=seq[i:].translate(standard_table)\n","  protein_pep.append(peptide)\n","# end at different nucleotide\n","for j in range (3,len(seq)):\n","  peptide=seq[:j].translate(standard_table)\n","  protein_pep.append(peptide)\n","\n","protein_pep = list(set(protein_pep))\n","protein_pep.sort(key=len, reverse=True)\n","print(protein_pep)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"Csy2DYp8Um8j","executionInfo":{"status":"ok","timestamp":1653252120035,"user_tz":-180,"elapsed":275,"user":{"displayName":"Svetlana Milrud","userId":"08332842915222559816"}},"outputId":"27fd7f0e-269b-4c06-84dc-d2f2e18ae6c8"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[Seq('STRAFYG'), Seq('VLEHSME'), Seq('Y*SILW'), Seq('STRAFY'), Seq('TRAFYG'), Seq('LEHSME'), Seq('STRAF'), Seq('EHSME'), Seq('*SILW'), Seq('RAFYG'), Seq('AFYG'), Seq('HSME'), Seq('SILW'), Seq('STRA'), Seq('STR'), Seq('ILW'), Seq('FYG'), Seq('SME'), Seq('ST'), Seq('ME'), Seq('YG'), Seq('LW'), Seq('W'), Seq('E'), Seq('S'), Seq('G')]\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/Bio/Seq.py:2983: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n","  BiopythonWarning,\n"]}]},{"cell_type":"markdown","source":["Second option\n","\n"],"metadata":{"id":"IzUyAuRkHtWk"}},{"cell_type":"code","source":["from Bio.Seq import Seq\n","from Bio.Data import CodonTable\n","\n","standard_table = CodonTable.unambiguous_dna_by_name[\"Standard\"]\n","protein_pep = []\n","seq = Seq(\"AGTACTAGAGCATTCTATGGAG\")\n","\n","# start from different nucleotide\n","for i in range (0,len(seq)-2):\n","  for j in range (2,len(seq)):\n","    peptide=seq[i:j].translate(standard_table)\n","    protein_pep.append(peptide)\n","\n","protein_pep = list(set(protein_pep))\n","protein_pep.sort(key=len, reverse=True)\n","print(protein_pep)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"vNRDesJvGeTS","executionInfo":{"status":"ok","timestamp":1653329507055,"user_tz":-180,"elapsed":260,"user":{"displayName":"Svetlana Milrud","userId":"08332842915222559816"}},"outputId":"c22b9adb-2101-4b76-8c57-b36d88a19016"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["[Seq('STRAFYG'), Seq('Y*SILW'), Seq('TRAFYG'), Seq('STRAFY'), Seq('VLEHSM'), Seq('STRAF'), Seq('TRAFY'), Seq('*SILW'), Seq('VLEHS'), Seq('Y*SIL'), Seq('RAFYG'), Seq('LEHSM'), Seq('VLEH'), Seq('LEHS'), Seq('RAFY'), Seq('AFYG'), Seq('SILW'), Seq('STRA'), Seq('Y*SI'), Seq('EHSM'), Seq('*SIL'), Seq('TRAF'), Seq('TRA'), Seq('Y*S'), Seq('STR'), Seq('AFY'), Seq('SIL'), Seq('FYG'), Seq('RAF'), Seq('EHS'), Seq('*SI'), Seq('LEH'), Seq('ILW'), Seq('HSM'), Seq('VLE'), Seq('RA'), Seq('HS'), Seq('LE'), Seq('IL'), Seq('SI'), Seq('YG'), Seq('SM'), Seq('VL'), Seq('LW'), Seq('Y*'), Seq('EH'), Seq('*S'), Seq('TR'), Seq('FY'), Seq('ST'), Seq('AF'), Seq('G'), Seq('I'), Seq('*'), Seq('W'), Seq('S'), Seq('T'), Seq('Y'), Seq('A'), Seq('L'), Seq('R'), Seq('F'), Seq('E'), Seq('V'), Seq('H'), Seq('M'), Seq('')]\n","67\n"]},{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.7/dist-packages/Bio/Seq.py:2983: BiopythonWarning: Partial codon, len(sequence) not a multiple of three. Explicitly trim the sequence or add trailing N before translation. This may become an error in future.\n","  BiopythonWarning,\n"]}]},{"cell_type":"markdown","source":["## Task 6\n","\n","TASK: Try to create one-line function (without (!!!) using Bio package)\n","that returns reverse complementary to a given sequence. \n","Hint: using dictionary & list comprehensions might be helpful."],"metadata":{"id":"vHlzDouzJxpA"}},{"cell_type":"code","execution_count":null,"metadata":{"id":"hO1jo4uETg3o"},"outputs":[],"source":["# for DNA\n","def rev_compl_one_line(seq):\n","    return (''.join(([{'A':'T','T':'A','C':'G','G':'C'}.get(x) for x in seq[::-1]])))"]},{"cell_type":"code","source":["# for RNA\n","def rev_compl_one_line_rna(seq):\n","    return (''.join(([{'A':'U','U':'A','C':'G','G':'C'}.get(x) for x in seq[::-1]])))"],"metadata":{"id":"TG2L_xd0I_mz"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["#DNA\n","seq='GAAACCCTTAC'\n","rev_compl_one_line(seq)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":36},"id":"T2-BcA6xHYAk","executionInfo":{"status":"ok","timestamp":1653248450914,"user_tz":-180,"elapsed":418,"user":{"displayName":"Svetlana Milrud","userId":"08332842915222559816"}},"outputId":"077b2241-b01a-40fb-fd93-7394579a8fa0"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["'GTAAGGGTTTC'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{},"execution_count":179}]},{"cell_type":"code","source":["#RNA\n","rna_seq='GAAACCCUUAC'\n","rev_compl_one_line_rna(rna_seq)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":36},"id":"faSn5QVjI4-J","executionInfo":{"status":"ok","timestamp":1653248448431,"user_tz":-180,"elapsed":289,"user":{"displayName":"Svetlana Milrud","userId":"08332842915222559816"}},"outputId":"9d63d068-61b8-40a1-c09e-1169b34a2552"},"execution_count":null,"outputs":[{"output_type":"execute_result","data":{"text/plain":["'GUAAGGGUUUC'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{},"execution_count":178}]}],"metadata":{"kernelspec":{"display_name":"Python 3","language":"python","name":"python3"},"language_info":{"codemirror_mode":{"name":"ipython","version":3},"file_extension":".py","mimetype":"text/x-python","name":"python","nbconvert_exporter":"python","pygments_lexer":"ipython3","version":"3.7.6"},"colab":{"name":"HW6_Milrud.ipynb","provenance":[]}},"nbformat":4,"nbformat_minor":0}